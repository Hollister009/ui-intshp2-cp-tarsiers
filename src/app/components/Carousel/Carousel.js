import React, { Component } from 'react';
import './_carousel.scss';
import ProductItem from '../ProductItem/ProductItem';

class Carousel extends Component {
  carouselStyle = {
    translation: 0,
    scrollCounter: 0
  };

  carouselStyleSheet = {
    transform: null
  };

  constructor(props) {
    super(props);
    this.state = {};
    this.wrapperRef = React.createRef();
  }

  updateTranslateStep = value => {
    this.carouselStyle = {
      ...this.carouselStyle,
      translateStep: value
    };
  };

  nextSlide = () => {
    const {
      translation,
      translateStep,
      scrollCounter,
      width
    } = this.carouselStyle;

    const wrapperWidth = this.wrapperRef.current.offsetWidth;

    this.carouselStyle = {
      ...this.carouselStyle,
      translation:
        translation -
        ((width + translation - wrapperWidth) / 2 <
        this.carouselStyle.translateStep
          ? width + translation - wrapperWidth
          : translateStep),
      scrollCounter: scrollCounter + 1
    };

    this.setState(state => ({
      ...state
    }));

    this.carouselStyleSheet = {
      ...this.carouselStyleSheet,
      transform: `translate(${this.carouselStyle.translation}px)`
    };
  };

  prevSlide = () => {
    const { translation, translateStep, scrollCounter } = this.carouselStyle;

    this.carouselStyle = {
      ...this.carouselStyle,
      translation:
        translation +
        (-translation / 2 < translateStep ? -translation : translateStep),
      scrollCounter: scrollCounter - 1
    };

    this.setState(state => ({
      ...state
    }));

    this.carouselStyleSheet = {
      ...this.carouselStyleSheet,
      transform: `translate(${this.carouselStyle.translation}px)`
    };
  };

  componentDidUpdate = () => {
    const { data } = this.props;

    this.carouselStyleSheet = {
      ...this.carouselStyleSheet,
      width: `${data.length * 300 - 30}px`
    };

    this.carouselStyle = {
      ...this.carouselStyle,
      width: data.length * 300 - 30
    };
  };

  render() {
    const { data, itemsPerView } = this.props;
    const { translation, scrollCounter } = this.carouselStyle;

    return (
      <div className="carousel--wrapper" ref={this.wrapperRef}>
        <div className="carousel" style={this.carouselStyleSheet}>
          {data.map(el => (
            <ProductItem
              key={el.id}
              updateTranslateStep={this.updateTranslateStep}
              data={el}
            />
          ))}
        </div>
        <button
          type="button"
          className="carousel--button carousel--button-prev"
          onClick={this.prevSlide}
          disabled={translation === 0}
        >
          <i className="fas fa-angle-left" />
        </button>
        <button
          type="button"
          className="carousel--button carousel--button-next"
          onClick={this.nextSlide}
          disabled={data.length - scrollCounter === itemsPerView}
        >
          <i className="fas fa-angle-right" />
        </button>
      </div>
    );
  }
}

export default Carousel;
