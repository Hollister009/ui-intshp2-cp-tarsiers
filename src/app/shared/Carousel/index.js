import React, { Component } from 'react';
import './Carousel.scss';

class Carousel extends Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.carouselRef = React.createRef();
    this.wrapperRef = React.createRef();
    this.isTouchDevice = 'ontouchstart' in window || navigator.msMaxTouchPoints;

    this.carouselStyle = {
      translation: 0,
      scrollCounter: 0,
      doubleSideMargin: 30
    };

    this.carouselStyleSheet = { transform: null };

    this.wrapperStyle = { overflowX: this.isTouchDevice ? 'scroll' : 'hidden' };
  }

  nextSlide = () => {
    const {
      translateStep,
      translation,
      scrollCounter,
      width
    } = this.carouselStyle;
    const wrapperWidth = this.wrapperRef.current.offsetWidth;

    this.carouselStyle = {
      ...this.carouselStyle,
      translation:
        translation -
        ((width + translation - wrapperWidth) / 2 < translateStep
          ? width + translation - wrapperWidth
          : translateStep),
      scrollCounter: scrollCounter + 1
    };

    this.setState(state => ({ ...state }));

    this.carouselStyleSheet = {
      ...this.carouselStyleSheet,
      transform: `translate(${this.carouselStyle.translation}px)`
    };
  };

  prevSlide = () => {
    const { translateStep, translation, scrollCounter } = this.carouselStyle;

    this.carouselStyle = {
      ...this.carouselStyle,
      translation:
        translation +
        (-translation / 2 < translateStep ? -translation : translateStep),
      scrollCounter: scrollCounter - 1
    };

    this.setState(state => ({ ...state }));

    this.carouselStyleSheet = {
      ...this.carouselStyleSheet,
      transform: `translate(${this.carouselStyle.translation}px)`
    };
  };

  componentDidUpdate = () => {
    const { data } = this.props;
    const { translateStep } = this.carouselStyle;

    this.carouselStyleSheet = {
      ...this.carouselStyleSheet,
      width: `${this.carouselRef.current.scrollWidth}px`
    };

    this.carouselStyle = {
      ...this.carouselStyle,
      width: this.carouselRef.current.scrollWidth,
      translateStep:
        (this.carouselRef.current.scrollWidth +
          this.carouselStyle.doubleSideMargin) /
        data.length,
      visibleItems: Math.ceil(
        this.wrapperRef.current.offsetWidth / translateStep
      )
    };
  };

  renderButtons() {
    const { translation, scrollCounter, visibleItems } = this.carouselStyle;
    const { data } = this.props;

    return this.isTouchDevice || visibleItems >= data.length ? null : (
      <React.Fragment>
        <button
          type="button"
          className="carousel__button carousel__button--prev"
          onClick={this.prevSlide}
          disabled={translation === 0}
        >
          <i className="fas fa-angle-left" />
        </button>

        <button
          type="button"
          className="carousel__button carousel__button--next"
          onClick={this.nextSlide}
          disabled={data.length - scrollCounter === visibleItems}
        >
          <i className="fas fa-angle-right" />
        </button>
      </React.Fragment>
    );
  }

  render() {
    const { children } = this.props;

    return (
      <div
        className="carousel__wrapper"
        style={this.wrapperStyle}
        ref={this.wrapperRef}
      >
        <div
          className="carousel"
          style={this.carouselStyleSheet}
          ref={this.carouselRef}
        >
          {children}
        </div>
        {this.renderButtons()}
      </div>
    );
  }
}

export default Carousel;
