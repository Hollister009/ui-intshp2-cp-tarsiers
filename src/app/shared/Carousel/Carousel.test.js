import React from 'react';
import { shallow } from 'enzyme';
import Carousel from './index';

const props = {
  children: {},
  data: []
};

describe('<Carousel />', () => {
  it('should match its snapshot', () => {
    const wrapper = shallow(<Carousel {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    expect(shallow(<Carousel {...props} />).length).toEqual(1);
  });
});
