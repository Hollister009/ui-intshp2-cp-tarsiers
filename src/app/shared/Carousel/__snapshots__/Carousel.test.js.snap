// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> should match its snapshot 1`] = `
<div
  className="carousel__wrapper"
  style={
    Object {
      "overflowX": "hidden",
    }
  }
>
  <div
    className="carousel"
    style={
      Object {
        "transform": null,
      }
    }
  >
    <Component />
  </div>
  <button
    className="carousel__button carousel__button--prev"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-angle-left"
    />
  </button>
  <button
    className="carousel__button carousel__button--next"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-angle-right"
    />
  </button>
</div>
`;
